CMP1903M Assessment 1 Base Code/Analyse.cs

```diff
@@ -1,54 +1,134 @@
+        public Dictionary<string, int> Characters { get; private set; }
+
+        public Analyse()
+        {
+            
+        }
+        /// <summary>
+        /// Runs analysis methods
+        /// </summary>
+        /// <param name="input">Text to analyse</param>
+        protected void AnalyseText(string input)
+        {
+            Elements = CountElements(input);
+            Characters = CountChars(input);
+        }
 
-        //Method: analyseText
-        //Arguments: string
-        //Returns: list of integers
-        //Calculates and returns an analysis of the text
-        public static Dictionary<string, int> AnalyseText(string input)
+        /// <summary>
+        /// Analyses text using instances of AnalysisTool
+        /// </summary>
+        /// <param name="text">Text to analyse</param>
+        /// <returns>dict: Keys as text elements and values corresponding to count</returns>
+        protected Dictionary<string, int> CountElements(string text)
         {
             //List of integers to hold the first five measurements:
             //1. Number of sentences
             //2. Number of vowels
             //3. Number of consonants
             //4. Number of upper case letters
             //5. Number of lower case letters
-            //Initialise all the values in the list to '0'
+            
+            // Create dictionary with anonymous functions that count elements using Regex
             var tools = new Dictionary<string, Func<string, int>>()
             {
-                {"sentences", s => Regex.Matches(s, @"(\s|^)+[^.!?]*([.!?]|\S)", RegexOptions.Multiline).Count()},
+                {"sentences", s => Regex.Matches(s, @"(\s|^)+[^.!?\*]*([.!?\*]|\S)", RegexOptions.Multiline).Count()},
                 {"vowels", s => Regex.Matches(s, @"[aeiouAEIOU]", RegexOptions.Multiline).Count()},
                 {"consonants", s => Regex.Matches(s, @"[^aeiouAEIOU\W\s\d]", RegexOptions.Multiline).Count()},
                 {"upper", s => Regex.Matches(s, @"[A-Z]", RegexOptions.Multiline).Count()},
                 {"lower", s => Regex.Matches(s, @"[a-z]", RegexOptions.Multiline).Count()}
             };
+            
+            // Return a dictionary. Use Linq to create a dictionary with elements as keys and returned values from
+            // instances of AnalysisTool as values
+            return tools.Keys.ToDictionary(toolsKey => toolsKey, toolsKey => new AnalysisTool(tools[toolsKey]).Count(text));
+        }
+        
+        /// <summary>
+        /// Counts number of characters in text
+        /// </summary>
+        /// <param name="text">Text to count characters in</param>
+        /// <returns>dict: Keys as characters and values corresponding to count</returns>
+        protected Dictionary<string, int> CountChars(string text)
+        {
+            var dict = new Dictionary<string, int>();
+            foreach (Match match in Regex.Matches(text.ToLower(), @"[a-z0-9]"))

```

Comment:
Brief asks for **letter** frequency, but feel free to leave the numbers in as well.
```suggestion
            foreach (Match match in Regex.Matches(text.ToLower(), @"[a-z]"))
```

Reply:
True, after some thought I'll accept the suggestion as it could cause some unexpected values during marking if only letters are specified. Thanks for pointing this out 👍